homeassistant/components/tasmota_irhvac/climate.py

```diff
@@ -0,0 +1,695 @@
+        self._enabled = False
+        self.power_mode = STATE_OFF
+        if initial_operation_mode is not STATE_OFF:
+            self.power_mode = STATE_ON
+            self._enabled = True
+        self._active = False
+        self._cur_temp = None
+        self._min_temp = min_temp
+        self._max_temp = max_temp
+        self._target_temp = target_temp
+        self._unit = hass.config.units.temperature_unit
+        self._support_flags = SUPPORT_FLAGS
+        if away_temp is not None:
+            self._support_flags = SUPPORT_FLAGS | SUPPORT_PRESET_MODE
+        self._away_temp = away_temp
+        self._is_away = False
+        self._modes_list = modes_list
+        self._fan_list = fan_list
+        self._swing_list = swing_list
+        self._quiet = quiet
+        self._turbo = turbo
+        self._econo = econo
+        self._model = model
+        self._celsius = celsius
+        self._light = light
+        self._filterr = filterr
+        self._clean = clean
+        self._beep = beep
+        self._sleep = sleep
+        self._sub_state = None
+
+    async def async_added_to_hass(self):
+        """Run when entity about to be added."""
+        await super().async_added_to_hass()
+        # Add listener
+        async_track_state_change(
+            self.hass, self.sensor_entity_id, self._async_sensor_changed
+        )
+        await self._subscribe_topics()
+
+        @callback
+        def _async_startup(event):
+            """Init on startup."""
+            sensor_state = self.hass.states.get(self.sensor_entity_id)
+            if sensor_state and sensor_state.state != STATE_UNKNOWN:
+                self._async_update_temp(sensor_state)
+
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, _async_startup)
+
+        # Check If we have an old state
+        old_state = await self.async_get_last_state()
+        if old_state is not None:
+            # If we have no initial temperature, restore
+            if old_state.attributes.get(ATTR_TEMPERATURE) is not None:
+                self._target_temp = float(old_state.attributes[ATTR_TEMPERATURE])
+            if old_state.attributes.get(ATTR_PRESET_MODE) == PRESET_AWAY:
+                self._is_away = True
+            if not self._hvac_mode and old_state.state:
+                self._hvac_mode = old_state.state
+                self._enabled = self._hvac_mode != STATE_OFF
+        else:
+            # No previous state, try and restore defaults
+            if self._target_temp is None:
+                self._target_temp = 26.0
+            _LOGGER.warning(
+                "No previously saved temperature, setting to %s", self._target_temp
+            )
+        # Set default state to off
+        if not self._hvac_mode:
+            self._hvac_mode = HVAC_MODE_OFF
+
+    async def _subscribe_topics(self):
+        """(Re)Subscribe to topics."""
+
+        @callback
+        def state_message_received(msg):
```

Comment:
Please extract the mqtt topic and payload parsing and subscription logic to a standalone library published on pypi.

See the mysensors integration for an example how to integrate and use the home assistant mqtt integration.

Reply:
Is your proposal mandatory or just an advice?

Reply:
Mandatory since it is in our [development checklist](https://developers.home-assistant.io/docs/en/development_checklist.html).

Reply:
Mandatory is only the part with the communication with other devices. As MQTT is the communication with other devices, it is separate part from my integration. I guess, you don't want to say that balllob and frenk are writing against their own rules, as the integrations that they are making and supporting are currently using `mqtt` in the same way. I don't see any of you stopped the `mqtt` integrations, like the `mqtt.switch` for example, or `manual_mqtt.alarm_control_panel`. I see Martin is very proud, that he has moved his mqtt requests outside of his `climate.py`, but this is his his style of codding in his situation with multiple communication protocols supported. Good job ,by the way. But I don't see any reason to accept Your advice for mandatory as more than the half of the integrations are currently written the same way, by the main developers of HA, like @balloob and @frenck  .. Please revise their integrations first, then revise mine again .. If You still think I have to create a separate class just to make one topic subscription and one topic publishing - no problem .. I can use my own integration as custom integration and I'll stop this pull request. But I'm interested in sharing it with everyone, as I believe it is useful. I think that you, guys, are understanding the mentioned above part of the check list in a different way. I'm not saying my integrations is the best in the world. I'm not even Python developer. I just decided to do the things the way that HA developers do it.

Reply:
Please follow the advice of Martin and don't tag Paulus or Franck. Mqtt is a transport protocol for what we use it.

Reply:
The integration and platforms you are shielding your statement with are implementing core platform components, your integration is not.

For integrations, we require logic to be abstracted into separate libraries. This has been done for a long time now and is documented. 

So I agree on the call Martin makes here since it fits.

Reply:
@frenck I'm sorry I passed the ball to you. I saw You in the contributors list for these platforms, and I know You're developing for HA, so thought You created these platforms.

@pvizeli As integrations count is too big, can you point me to another integration, that has its use of mqtt outside of the main integration file? I want to see if there is a simple way of doing things. Martin's code it too complicated and best suits the use of multiple protocols. I feel I need something simpler and cleaner.

As for the other part, I feel very proud of myself for developing core platform component. :D

Reply:
In basic terms, you want to pass callbacks to your library so that the library can hook into the home assistant mqtt integration. Here's the example from mysensors for that:

```py
mqtt = hass.components.mqtt

def pub_callback(topic, payload, qos, retain):
    """Call MQTT publish function."""
    mqtt.async_publish(topic, payload, qos, retain)

def sub_callback(topic, sub_cb, qos):
    """Call MQTT subscribe function."""

    @callback
    def internal_callback(msg):
        """Call callback."""
        sub_cb(msg.topic, msg.payload, msg.qos)

    hass.async_create_task(mqtt.async_subscribe(topic, internal_callback, qos))
```

This example uses the async interface so needs to be run within the event loop also in the library. If the library will not support asyncio we have to change to use the home assistant sync interface.

Reply:
Here's a sync example:
```py
mqtt = hass.components.mqtt

def pub_callback(topic, payload, qos, retain):
    """Call MQTT publish function."""
    mqtt.publish(topic, payload, qos, retain)

def sub_callback(topic, sub_cb, qos):
    """Call MQTT subscribe function."""

    def internal_callback(msg):
        """Call callback."""
        sub_cb(msg.topic, msg.payload, msg.qos)

    mqtt.subscribe(topic, internal_callback, qos)
```